[collector] Add support for NetworkPolicy

fernferret opened the pull request:
Adds NetworkPolicy support to the collector chart.

Fixes #722

## Questions for reviewers

~~When `networkPolicy` is enabled, I decided to include a default policy of "all namespaces".~~

Per the [discussion below](https://github.com/open-telemetry/opentelemetry-helm-charts/pull/723#issuecomment-1495841280), when `networkPolicy` is enabled with no other arguments, a `NetworkPolicy` object will be created, but traffic will not be restricted until `allowIngressFrom` is set to a `from` clause of a network policy

```yaml
networkPolicy:
  enabled: true
```

Here is an example that would restrict traffic to only allow ingress from pods in the cluster:

```yaml
networkPolicy:
  enabled: true
  allowIngressFrom:
    # Allow traffic from any pod in any namespace, but not external hosts
    - namespaceSelector: {}
```

fernferret commented:
Whoopsie, fixed the whitespace issues... not sure how those crept in. My local run of `ct` passes now:

```
ct lint --target-branch main --chart-repos opentelemetry-collector=https://open-telemetry.github.io/opentelemetry-helm-charts --chart-repos prometheus=https://prometheus-community.github.io/helm-charts --chart-repos grafana=https://grafana.github.io/helm-charts --chart-repos jaeger=https://jaegertracing.github.io/helm-chart
...
1 chart(s) linted, 0 chart(s) failed
------------------------------------------------------------------------------------------------------------------------
 ✔︎ opentelemetry-collector => (version: "0.52.0", path: "charts/opentelemetry-collector")
------------------------------------------------------------------------------------------------------------------------
All charts linted successfully
```

TylerHelmuth commented on charts/opentelemetry-collector/templates/networkpolicy.yaml:
```diff
@@ -0,0 +1,39 @@
+  egress:
+    {{- toYaml .Values.networkPolicy.egressRules | nindent 4 }}
+  {{- end }}
+{{- end }}
```
```suggestion
{{- end }}

```

TylerHelmuth commented on charts/opentelemetry-collector/templates/networkpolicy.yaml:
```diff
@@ -0,0 +1,39 @@
+    - ports:
+        {{- range $key, $port := .Values.ports }}
+        {{- if $port.enabled }}
+        # {{ $key }}
```
Can we remove this comment if not needed?

fernferret replied:
Done, I also removed the `$key` as that's just extra if the comment is gone.

TylerHelmuth reviewed (commented):
I think the decision you laid out in the description makes sense, thanks for the thoughtful write-up.  

Can you add a new test scenario to the CI folder that enables this new capability?

Please rebase and then run `make generate-examples CHARTS=opentelemetry-collector`.

fernferret commented:
@TylerHelmuth Thank you for the feedback. These are all great suggestions. I have the patches made and rebased and will push very soon.

Before I get this pushed, I did have one question about what you asked for:

> Can you add a new test scenario to the CI folder that enables this new capability?

Is there something else to do other than add new `values.yaml` files to the `ci` dir? I added 2 new files (), one to showcase the default capability, and one with a bit more customization. 

I've not used `ct` before, is there any guidance or documentation on how I could add these or verify the tests are working?

TylerHelmuth commented:
@fernferret Adding new files, one per test scenario, is all that's needed.

fernferret commented on charts/opentelemetry-collector/ci/networkpolicy-override-values.yaml:
@TylerHelmuth I apologize for my confusion, but are these the `ci` files you requested be added? I believe these were part of the original PR. If they're not I'm happy to change/move them.

TylerHelmuth replied:
I'm not sure what you mean by "original PR", but yes, these are the test scenarios I'm looking for.

TylerHelmuth replied:
Oh were these part of the first commit and I missed them?

fernferret replied:
> Oh were these part of the first commit and I missed them?

I believe they were, no worries at all I just wanted to make sure I was following all the feedback! I'll get the rebased/`make`'d changes pushed soon!

TylerHelmuth commented:
@povilasv I actually like the idea that we can provide a default policy of all kubernetes pods/hosts and exclude external traffic.  For the collector, especially daemonset mode, its reasonable to expect that the only thing communicating with it is traffic from inside the cluster.  The default value is also very easy to override.  

All that said, I don't have experience using NetworkPolicy myself. If it isn't standard for a chart to provide a default value the we could remove it.  @dmitryax what do you think?

fernferret commented:
After diving a bit deeper, I agree with @povilasv and think the default values file should contain the following. I also like providing reasonable alternatives as commented out examples.

```yaml
networkPolicy:
  enabled: false

  # Annotations to add to the NetworkPolicy
  annotations: {}

  # Configure the 'from' clause of the NetworkPolicy
  # By default this will restrict traffic to the listed ports only.
  # If you wish to further restrict traffic to other hosts or specific
  # namespaces, see the standard NetworkPolicy 'spec.ingress.from'
  # definition for more info:
  # https://kubernetes.io/docs/reference/kubernetes-api/policy-resources/network-policy-v1/
  allowIngressFrom: []
  # Allow traffic from any pod in any namespace, but not external hosts
  # - namespaceSelector: {}
```

---

There's been some really good discussion and feedback from both of y'all that's caused me to really dig into `NetworkPolicy`.  I'm using them to restrict tenant access to the `DaemonSet`, but I definitely want to make sure this PR breaks as little as possible and works for people!

## About NetworkPolicy

The following 3 bullets (extracted for readability) come from the [K8S API describing the `NetworkPolicy`'s `spec.ingress.from`](https://kubernetes.io/docs/reference/kubernetes-api/policy-resources/network-policy-v1/#NetworkPolicySpec) which when combined paint a much better picture:

* If this field (`spec.ingress.from`) is empty or missing, this rule matches all sources (traffic not restricted by source).
* If this field (`spec.ingress.from`) is present and contains at least on item, this rule allows traffic only if the traffic matches at least one item in the from list.
* If this field (`spec.ingress.ports`) is present and contains at least one item, then this rule allows traffic only if the traffic matches at least one port in the list.

So, given the following policy:

```yaml
spec:
  ingress:
  - ports:
    - port: 4318
      protocol: TCP
    - port: 4317
      protocol: TCP
```

I believe this means that both a pod and an external user, would be able to access the collector on only ports `4317` and `4318`. However, if a malicious actor somehow figured out how to start up another application on another port (let's say `1234`) that port would not be accessible inside or out.

This also means that external pods won't be able to hit the health port `13113` unless it is exposed (which is available to advanced users with `extraIngressRules`), however I think this makes for a sensible default as `13113` is not bound to the Pod IP as far as I see.

Since we're also dealing with `hostPort`s I'd also like the ability to restrict things further (as described in the PR/issue description).

This example would allow only traffic from inside the cluster, thus allowing all pods within to talk to the collector, but not my development laptop, even in DaemonSet mode:

```yaml
spec:
  ingress:
  - ports:
    - port: 4318
      protocol: TCP
    - port: 4317
      protocol: TCP
    from:
    - namespaceSelector: {}
```

And finally, if someone wanted to allow all pods within the cluster **and** a subset of development nodes, which is my use case:

```yaml
spec:
  ingress:
  - ports:
    - port: 4318
      protocol: TCP
    - port: 4317
      protocol: TCP
    from:
    - namespaceSelector: {}
    - ipBlock:
        cidr: 192.168.1.0/24
```

After thinking through and reviewing these 3 use cases, I think the first one, with no default `from` clause as @povilasv suggested, makes the most sense as it does provide some guarantees that no other ports will be exposed.

## Relevant Charts

I looked at the prometheus chart and it actually has a reasonably stock [NetworkPolicy](https://github.com/prometheus-community/helm-charts/blob/main/charts/prometheus/templates/network-policy.yaml).

I followed is lineage back to: https://github.com/prometheus-community/helm-charts/commit/ac9c25afcd2268897d586ddbae39f36b82240192.

I would certainly like to see some more options added to the Prometheus chart, but it does point at the fact that:

* network policy came from a place where some portions of it were defined with annotations
* for many deployments, network policy is targeted at ensuring only the known ports are exposed

## Test results

Here are some results from my kubeadm provisioned dev cluster, just so I could verify my own understanding.

<details>
<summary>Show Test Results</summary>

### No NetworkPolicy

Allow traffic from all pods and hosts, to all ports

Ports active: `4318`, `4317`

```
$ curl --connect-timeout 5 http://farnsworth:4318
404 page not found
$ curl --connect-timeout 5 http://farnsworth:4317
curl: (1) Received HTTP/0.9 when not allowed       <-- traffic made it to the pod
```

### Empty NetworkPolicy `from` clause

Allow traffic from all pods and hosts, but only to 4317

Ports active: `4318`, `4317`

```yaml
spec:
  ingress:
  - ports:
    - port: 4318
      protocol: TCP
```

```
$ curl --connect-timeout 5 http://farnsworth:4318
404 page not found
$ curl --connect-timeout 5 http://farnsworth:4317
curl: (28) Connection timeout after 5002 ms
```

### Only allow internal traffic

Allow traffic from all pods to 4318, but no external hosts. No traffic allowed to 4317.

Ports active: `4318`, `4317`

```yaml
spec:
  ingress:
  - ports:
    - port: 4318
      protocol: TCP
    from:
    - namespaceSelector: {}
```

```
$ curl --connect-timeout 5 http://farnsworth:4318
curl: (28) Connection timeout after 5004 ms
$ curl --connect-timeout 5 http://farnsworth:4317
curl: (28) Connection timeout after 5004 ms
```

### Allow internal + development laptop

Allow traffic from all pods and hosts to 4318. No traffic allowed to 4317.

Ports active: `4318`, `4317`

```yaml
spec:
  ingress:
  - ports:
    - port: 4318
      protocol: TCP
    from:
    - namespaceSelector: {}
    - ipBlock:
        cidr: 192.168.1.0/24
```

```
$ curl --connect-timeout 5 http://farnsworth:4318
404 page not found
$ curl --connect-timeout 5 http://farnsworth:4317
curl: (28) Connection timeout after 5003 ms
```

</details>

## Wrap Up

I appreciate y'all continuing the discussion and look forward to your advice.

fernferret commented on charts/opentelemetry-collector/values.yaml:
```diff
@@ -432,3 +432,56 @@ statefulset:
+  annotations: {}
+
+  # Configure the 'from' clause of the NetworkPolicy
+  # By default this has no effect in Deployment or StatefulSet modes
+  # In DaemonSet mode, this will restrict access to pods (no external hosts)
```
Thanks for pointing this out, I really dislike how I phrased this and have proposed a change in the main comment thread.

fernferret replied:
This comment has been removed.

TylerHelmuth commented:
@fernferret thanks again for the beautiful writes, it makes reviewing and discussing a new feature so much easier.  I wish all the PRs I reviewed had this level of investment and presentation haha

Your extra digging and explain, along with Prometheus as an example, feel like a good argument for not providing any default `from`.  I support the change suggested by you and @povilasv.

fernferret commented:
I appreciate the feedback!

I've made the requested changes (as well as reworded the comment block within the `networkPolicy` section) and they're pushed for review.

TylerHelmuth commented:
@fernferret needs one more rebase (I merged the other PR first)

puckpuck commented:
Would network policy allow the collector to work with `kubectl port-forward ...` or would we still need to force it to bind to 127.0.0.1?
